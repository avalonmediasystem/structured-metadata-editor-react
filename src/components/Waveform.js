import React, { Component } from 'react';
import Peaks from 'peaks.js';
import {
	Button,
	ButtonToolbar,
	FormControl,
	FormGroup,
	Form,
	Row,
	Col
} from 'react-bootstrap';

import soundMP3 from '../data/TOL_6min_720p_download.mp3';
import soundOGG from '../data/TOL_6min_720p_download.ogg';

const peaksOptions = {
	container: null,
	mediaElement: null,
	dataUri: '../data/TOL_6min_720p_download.json',
	dataUriDefaultFormat: 'json',
	keyboard: true,
	pointMarkerColor: '#006eb0',
	showPlayheadTime: true
};

class Waveform extends Component {
	constructor(props) {
		super(props);
		this.state = {
			seekTime: ''
		};

		this.peaksInstance = null;

		// Create refs here
		this.waveformContainer = React.createRef();
		this.mediaPlayer = React.createRef();

		this.handleChange = this.handleChange.bind(this);
	}

	componentDidMount() {
		// Grab the React `refs` now that the component has mounted
		peaksOptions.container = this.waveformContainer.current;
		peaksOptions.mediaElement = this.mediaPlayer.current;

		// Initialize Peaks
		this.peaksInstance = Peaks.init(peaksOptions);
	}

	zoomIn = () => {
		this.peaksInstance.zoom.zoomIn();
	};

	zoomOut = () => {
		this.peaksInstance.zoom.zoomOut();
	};

	handleChange(event) {
		this.setState({
			seekTime: event.target.value
		});
	}

	seekTime = () => {
		const timeInSeconds = parseFloat(this.state.seekTime);
		if (!Number.isNaN(timeInSeconds)) {
			this.peaksInstance.player.seek(timeInSeconds);
		}
	};

	render() {
		return (
			<section className="waveform-section">
				<div id="waveform-container" ref={this.waveformContainer} />
				<Row>
					<Col xs={12} md={6}>
						<audio controls ref={this.mediaPlayer}>
							<source src={soundMP3} type="audio/mp3" />
							<source src={soundOGG} type="audio/ogg" />
							Your browser does not support the audio element.
						</audio>
					</Col>
					<Col xs={12} md={6} className="text-right">
						<Form inline>
							<FormGroup>
								<ButtonToolbar>
									<Button onClick={this.zoomIn}>Zoom in</Button>
									<Button onClick={this.zoomOut}>Zoom out</Button>
								</ButtonToolbar>
							</FormGroup>{' '}
							<FormControl
								className="form-control"
								type="text"
								value={this.state.seekTime}
								onChange={this.handleChange}
								placeholder="0"
							/>{' '}
							<Button onClick={this.seekTime}>Seek</Button>
						</Form>
					</Col>
				</Row>
			</section>
		);
	}
}

export default Waveform;
