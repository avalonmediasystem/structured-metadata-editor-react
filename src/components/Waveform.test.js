import React from 'react';
import Waveform, { PureWaveform } from '../components/Waveform';
import { shallow, mount } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import Peaks from 'peaks';

const mockStore = configureMockStore([thunk]);

describe('Waveform component', () => {
  let store, peaks;
  let options = {
    container: null,
    mediaElement: null,
    dataUri: null,
    dataUriDefaultFormat: 'json',
    keyboard: true,
    _zoomLevelIndex: 0,
    _zoomLevels: [512, 1024, 2048, 4096]
  };
  beforeEach(() => {
    peaks = Peaks.init(options);
    store = mockStore({
      peaksInstance: {
        peaks: peaks,
        events: undefined
      }
    });
  });

  test('renders Waveform without crashing', () => {
    const waveformContainer = React.createRef();
    const mediaPlayer = React.createRef();
    const wrapper = shallow(
      <Provider store={store}>
        <Waveform ref={waveformContainer} ref={mediaPlayer} />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('#waveform-container')).toBeDefined();
    expect(wrapper.find('audio')).toBeDefined();
  });

  describe('tests Waveform component view', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(
        <Provider store={store}>
          <Waveform waveformRef={() => {}} mediaPlayerRef={() => {}} />
        </Provider>
      );
    });

    test('renders audio element with src attribute', () => {
      expect(wrapper.find('audio').instance().src).toBe(
        'http://localhost/utah_phillips_one.mp3'
      );
      expect(wrapper.find('audio').instance().controls).toBeTruthy();
    });

    test('renders props correctly', () => {
      expect(
        wrapper.find(Waveform).instance().props.mediaPlayerRef
      ).toBeDefined();
      expect(wrapper.find(Waveform).instance().props.waveformRef).toBeDefined();
    });

    test('renders waveform control buttons', () => {
      expect(
        wrapper
          .find('Button')
          .at(0)
          .instance().props.className
      ).toBe('glyphicon glyphicon-zoom-in');

      expect(
        wrapper
          .find('Button')
          .at(1)
          .instance().props.className
      ).toBe('glyphicon glyphicon-zoom-out');
    });
  });

  test('tests zoom in button click', () => {
    const wrapper = mount(
      <Provider store={store}>
        <Waveform waveformRef={() => {}} mediaPlayerRef={() => {}} />
      </Provider>
    );

    expect(
      wrapper.instance().props.store.getState().peaksInstance.peaks.zoom
        ._zoomLevelIndex
    ).toEqual(2);

    wrapper
      .find('Button')
      .at(0)
      .simulate('click');

    expect(
      wrapper.instance().props.store.getState().peaksInstance.peaks.zoom
        ._zoomLevelIndex
    ).toEqual(1);

    expect(
      wrapper.instance().props.store.getState().peaksInstance.peaks.zoom.zoomIn
    ).toHaveBeenCalledTimes(1);
  });

  test('tests zoom out button click', () => {
    const wrapper = mount(
      <Provider store={store}>
        <Waveform waveformRef={() => {}} mediaPlayerRef={() => {}} />
      </Provider>
    );

    expect(
      wrapper.instance().props.store.getState().peaksInstance.peaks.zoom
        ._zoomLevelIndex
    ).toEqual(2);

    wrapper
      .find('Button')
      .at(1)
      .simulate('click');

    expect(
      wrapper.instance().props.store.getState().peaksInstance.peaks.zoom
        ._zoomLevelIndex
    ).toEqual(3);

    expect(
      wrapper.instance().props.store.getState().peaksInstance.peaks.zoom.zoomOut
    ).toHaveBeenCalledTimes(1);
  });

  test('tests input form for change event', () => {
    const wrapper = mount(
      <PureWaveform
        waveformRef={() => {}}
        mediaPlayerRef={() => {}}
        peaksInstance={peaks}
      />
    );

    expect(wrapper.find('FormControl').instance().props.value).toEqual('');

    // Mock user input through form
    wrapper.find('FormControl').simulate('change', { target: { value: '36' } });

    expect(wrapper.find('FormControl').instance().props.value).toEqual('36');
  });

  test('tests seek time button click', () => {
    let peaksObj = { peaks: peaks };
    const wrapper = mount(
      <PureWaveform
        waveformRef={() => {}}
        mediaPlayerRef={() => {}}
        peaksInstance={peaksObj}
      />
    );

    // Update state
    wrapper.setState({ seekTime: '36' });

    // Test current time of player instance before clicking seek button
    expect(
      wrapper.instance().props.peaksInstance.peaks.player._mediaElement
        .currentTime
    ).toEqual(0);

    wrapper
      .find('Button')
      .at(2)
      .simulate('click');

    // Test current time of the player instance after seek button is clicked
    expect(
      wrapper.instance().props.peaksInstance.peaks.player._mediaElement
        .currentTime
    ).toEqual(36);
  });
});
