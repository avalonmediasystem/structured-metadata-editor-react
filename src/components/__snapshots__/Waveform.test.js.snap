// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Waveform component renders Waveform without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <Connect(Waveform) />
  </Provider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Connect(Waveform) />,
      "value": Object {
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeState": Object {
          "peaksInstance": Object {
            "peaks": Object {
              "options": Object {
                "_zoomLevelIndex": 0,
                "_zoomLevels": Array [
                  512,
                  1024,
                  2048,
                  4096,
                ],
                "container": null,
                "dataUri": null,
                "dataUriDefaultFormat": "json",
                "keyboard": true,
                "mediaElement": null,
              },
              "player": Object {
                "_mediaElement": Object {
                  "currentTime": 0,
                },
                "seek": [MockFunction],
              },
              "segments": Object {
                "_peaks": [Circular],
                "_segments": Array [
                  Object {
                    "color": "#80A590",
                    "endTime": 360,
                    "id": "123a-456b-789c-2d",
                    "labelText": "Sample segment",
                    "startTime": 0,
                  },
                  Object {
                    "color": "#2A5459",
                    "endTime": 1259.99,
                    "id": "123a-456b-789c-9d",
                    "labelText": "Last segment",
                    "startTime": 750,
                  },
                ],
                "add": [MockFunction],
                "getSegment": [MockFunction],
                "getSegments": [MockFunction],
                "removeAll": [MockFunction],
                "removeById": [MockFunction],
              },
              "zoom": Object {
                "_peaks": [Circular],
                "_zoomLevelIndex": 2,
                "_zoomLevels": Array [
                  512,
                  1024,
                  2048,
                  4096,
                ],
                "zoomIn": [MockFunction],
                "zoomOut": [MockFunction],
              },
            },
          },
        },
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {},
      "ref": Object {
        "current": null,
      },
      "rendered": null,
      "type": [Function],
    },
    "type": Object {
      "$$typeof": Symbol(react.provider),
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "_calculateChangedBits": null,
          "_context": [Circular],
        },
        "Provider": [Circular],
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": null,
        "_currentValue2": null,
        "_threadCount": 0,
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Connect(Waveform) />,
        "value": Object {
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "storeState": Object {
            "peaksInstance": Object {
              "peaks": Object {
                "options": Object {
                  "_zoomLevelIndex": 0,
                  "_zoomLevels": Array [
                    512,
                    1024,
                    2048,
                    4096,
                  ],
                  "container": null,
                  "dataUri": null,
                  "dataUriDefaultFormat": "json",
                  "keyboard": true,
                  "mediaElement": null,
                },
                "player": Object {
                  "_mediaElement": Object {
                    "currentTime": 0,
                  },
                  "seek": [MockFunction],
                },
                "segments": Object {
                  "_peaks": [Circular],
                  "_segments": Array [
                    Object {
                      "color": "#80A590",
                      "endTime": 360,
                      "id": "123a-456b-789c-2d",
                      "labelText": "Sample segment",
                      "startTime": 0,
                    },
                    Object {
                      "color": "#2A5459",
                      "endTime": 1259.99,
                      "id": "123a-456b-789c-9d",
                      "labelText": "Last segment",
                      "startTime": 750,
                    },
                  ],
                  "add": [MockFunction],
                  "getSegment": [MockFunction],
                  "getSegments": [MockFunction],
                  "removeAll": [MockFunction],
                  "removeById": [MockFunction],
                },
                "zoom": Object {
                  "_peaks": [Circular],
                  "_zoomLevelIndex": 2,
                  "_zoomLevels": Array [
                    512,
                    1024,
                    2048,
                    4096,
                  ],
                  "zoomIn": [MockFunction],
                  "zoomOut": [MockFunction],
                },
              },
            },
          },
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": Object {
          "current": null,
        },
        "rendered": null,
        "type": [Function],
      },
      "type": Object {
        "$$typeof": Symbol(react.provider),
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": Object {
            "$$typeof": Symbol(react.context),
            "_calculateChangedBits": null,
            "_context": [Circular],
          },
          "Provider": [Circular],
          "_calculateChangedBits": null,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": null,
          "_currentValue2": null,
          "_threadCount": 0,
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
